packages/meditrak-server/src/kobo/startSyncWithKoBo.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import keyBy from 'lodash.keyby';
+import winston from 'winston';
+import { DataBroker } from '@tupaia/data-broker';
+import { generateId } from '@tupaia/database';
+import { reformatDateStringWithoutTz } from '@tupaia/utils';
+
+const PERIOD_BETWEEN_SYNCS = 10 * 60 * 1000; // 10 minutes between syncs
+const SERVICE_NAME = 'KoBo';
+
+export async function startSyncWithKoBo(models) {
+  if (process.env.KOBO_SYNC_DISABLE) {
+    // eslint-disable-next-line no-console
+    console.log('KoBo sync is disabled');
+  } else {
+    const dataBroker = new DataBroker();
+    setInterval(() => syncWithKoBo(models, dataBroker), PERIOD_BETWEEN_SYNCS);
+  }
+}
+
+export async function syncWithKoBo(models, dataBroker) {
+  await pullLatest(models, dataBroker);
+}
+
+async function pullLatest(models, dataBroker) {
+  const syncCursor = await models.syncCursor.findOne({ service_name: SERVICE_NAME });
+
+  // Pull data from KoBo
+  const koboData = await dataBroker.pull(
+    {
+      code: syncCursor.config.koboSurveys,
+      type: dataBroker.getDataSourceTypes().SYNC_GROUP,
+    },
+    {
+      startSubmissionTime: reformatDateStringWithoutTz(syncCursor.sync_time),
+    },
+  );
+
+  const apiUser = await models.apiClient.findOne({ username: process.env.KOBO_API_USERNAME });
+  if (!apiUser) {
+    throw new Error('Cannot find API client user for KoBo');
+  }
+
+  // Create new survey_responses
+  let newSyncTime = syncCursor.sync_time;
+  await models.wrapInTransaction(async transactingModels => {
+    for (const [surveyCode, responses] of Object.entries(koboData)) {
+      const survey = await transactingModels.survey.findOne({ code: surveyCode });
+
+      for (const responseData of responses) {
+        if (responseData.eventDate > newSyncTime) {
+          newSyncTime = responseData.eventDate;
+        }
+        const entity = await transactingModels.entity.findOne({ code: responseData.orgUnit });
+        if (!entity) {
+          winston.warn(`Skipping KoBo sync for unknown entity ${responseData.orgUnit}`);
```

Comment:
Currently just skipping responses if we can't map them to an entity. I'm not sure if this is the behaviour we want, open to alternatives

Reply:
I guess the other option is to block sync until it gets deleted or the entity gets mapped correctly. I think this is probably better :+1: